src/fsrs/impl/long_term_schduler.ts

```diff
@@ -0,0 +1,279 @@
+    const next_again = TypeConvert.card(this.current)
+    const next_hard = TypeConvert.card(this.current)
+    const next_good = TypeConvert.card(this.current)
+    const next_easy = TypeConvert.card(this.current)
+
+    this.next_ds(
+      next_again,
+      next_hard,
+      next_good,
+      next_easy,
+      difficulty,
+      stability,
+      retrievability
+    )
+
+    this.next_interval(next_again, next_hard, next_good, next_easy, interval)
+    this.next_state(next_again, next_hard, next_good, next_easy)
+    next_again.lapses += 1
+
+    this.update_next(next_again, next_hard, next_good, next_easy)
+    return this.next.get(grade)!
+  }
+
+  /**
+   * Review next_ds
+   */
+  private next_ds(
+    next_again: Card,
+    next_hard: Card,
+    next_good: Card,
+    next_easy: Card,
+    difficulty: number,
+    stability: number,
+    retrievability: number
+  ): void {
+    next_again.difficulty = this.algorithm.next_difficulty(
+      difficulty,
+      Rating.Again
+    )
+    next_again.stability = this.algorithm.next_forget_stability(
+      difficulty,
+      stability,
+      retrievability
+    )
+    next_hard.difficulty = this.algorithm.next_difficulty(
+      difficulty,
+      Rating.Hard
+    )
+    next_hard.stability = this.algorithm.next_recall_stability(
+      difficulty,
+      stability,
+      retrievability,
+      Rating.Hard
+    )
+    next_good.difficulty = this.algorithm.next_difficulty(
+      difficulty,
+      Rating.Good
+    )
+    next_good.stability = this.algorithm.next_recall_stability(
+      difficulty,
+      stability,
+      retrievability,
+      Rating.Good
+    )
+    next_easy.difficulty = this.algorithm.next_difficulty(
+      difficulty,
+      Rating.Easy
+    )
+    next_easy.stability = this.algorithm.next_recall_stability(
+      difficulty,
+      stability,
+      retrievability,
+      Rating.Easy
+    )
+  }
+
+  /**
+   * Review/New next_interval
+   */
+  private next_interval(
+    next_again: Card,
+    next_hard: Card,
+    next_good: Card,
+    next_easy: Card,
+    interval: number
+  ): void {
+    let again_interval: int,
+      hard_interval: int,
+      good_interval: int,
+      easy_interval: int
+    again_interval = this.algorithm.next_interval(
+      next_again.stability,
+      interval
+    )
+    hard_interval = this.algorithm.next_interval(next_hard.stability, interval)
+    good_interval = this.algorithm.next_interval(next_good.stability, interval)
+    easy_interval = this.algorithm.next_interval(next_easy.stability, interval)
+
+    again_interval = Math.min(again_interval, hard_interval) as int
+    hard_interval = Math.max(hard_interval, again_interval + 1) as int
+    good_interval = Math.max(good_interval, hard_interval + 1) as int
+    easy_interval = Math.max(easy_interval, good_interval + 1) as int
+
+    next_again.scheduled_days = again_interval
+    next_again.due =
+      again_interval > 0
```

Comment:
When would it be less or equal to zero?

Reply:
> When would it be less or equal to zero?

I have reviewed the code and confirmed that no cases of <=0 exist; this was my mistake.

https://github.com/open-spaced-repetition/ts-fsrs/blob/a752a6e172fb18879e710d008d5fbb4949bb8730/src/fsrs/algorithm.ts#L154-L164